### ISSUE TITLE
Implement MkDocs Material Documentation with Custom Dracula-Scientific Theme

### ISSUE THREAD

**Issue by Anselmoo** [Anselmoo/useful-optimizer#48]
## Summary

Implement comprehensive scientific documentation using **MkDocs Material** with a custom **Dracula-inspired color scheme** that emphasizes the scientific importance and research credibility of the useful-optimizer library.

## Decision Rationale

After thorough analysis of MkDocs Material, Docusaurus, and VitePress:
- **MkDocs Material** selected for: Python-native toolchain, Vega-Lite JSON charts (no PNG storage), superior LaTeX/math support, lower maintenance cost
- **Dracula palette** from [anselmoo.github.io/dracula-palette](https://anselmoo.github.io/dracula-palette/) for distinctive scientific branding
- **JSON-based charts** via `mkdocs-charts-plugin` to avoid Git LFS/PNG storage issues

---

### Color Palette (from Dracula)

| Role | Hex Code | Usage |
|------|----------|-------|
| **Background** | `#282a36` | Dark mode background |
| **Current Line** | `#44475a` | Code block background, cards |
| **Selection** | `#6272a4` | Hover states, secondary elements |
| **Foreground** | `#f8f8f2` | Primary text |
| **Comment** | `#6272a4` | Secondary text, captions |
| **Cyan** | `#8be9fd` | Primary accent (links, buttons) |
| **Green** | `#50fa7b` | Success states, positive values |
| **Orange** | `#ffb86c` | Warnings, highlights |
| **Pink** | `#ff79c6` | Secondary accent |
| **Purple** | `#bd93f9` | **Primary brand color**, headers |
| **Red** | `#ff5555` | Errors, critical notes |
| **Yellow** | `#f1fa8c` | Tips, important callouts |

### CSS Variables Implementation

```css
/* docs/stylesheets/dracula-scientific.css */

/* ===== DARK MODE (Primary) ===== */
[data-md-color-scheme="dracula"] {
  /* Core Dracula colors */
  --md-primary-fg-color: #bd93f9;        /* Purple - brand */
  --md-primary-fg-color--light: #caa9fa;
  --md-primary-fg-color--dark: #9d79d9;
  
  --md-accent-fg-color: #8be9fd;         /* Cyan - interactive */
  --md-accent-fg-color--transparent: rgba(139, 233, 253, 0.1);
  
  /* Background hierarchy */
  --md-default-bg-color: #282a36;
  --md-default-bg-color--light: #44475a;
  --md-default-bg-color--lighter: #6272a4;
  --md-default-bg-color--lightest: #7c8db5;
  
  /* Text colors */
  --md-default-fg-color: #f8f8f2;
  --md-default-fg-color--light: rgba(248, 248, 242, 0.7);
  --md-default-fg-color--lighter: rgba(248, 248, 242, 0.5);
  --md-default-fg-color--lightest: rgba(248, 248, 242, 0.3);
  
  /* Code blocks */
  --md-code-bg-color: #44475a;
  --md-code-fg-color: #f8f8f2;
  --md-code-hl-color: rgba(139, 233, 253, 0.2);
  
  /* Admonition colors */
  --md-admonition-bg-color: #44475a;
  
  /* Typeset enhancement */
  --md-typeset-a-color: #8be9fd;
  
  /* Footer */
  --md-footer-bg-color: #21222c;
  --md-footer-bg-color--dark: #191a21;
}

/* ===== LIGHT MODE (Alternative) ===== */
[data-md-color-scheme="dracula-light"] {
  --md-primary-fg-color: #6d44b8;        /* Darker purple for contrast */
  --md-primary-fg-color--light: #8b6bc4;
  --md-primary-fg-color--dark: #5533a0;
  
  --md-accent-fg-color: #0891b2;         /* Darker cyan */
  
  --md-default-bg-color: #f8f8f2;
  --md-default-bg-color--light: #e8e8e8;
  
  --md-default-fg-color: #282a36;
  --md-default-fg-color--light: rgba(40, 42, 54, 0.7);
  
  --md-code-bg-color: #e8e8e8;
  --md-code-fg-color: #282a36;
}

/* ===== SCIENTIFIC TYPOGRAPHY ===== */
:root {
  /* Headers - scientific gravitas */
  --md-text-font: "Inter", -apple-system, BlinkMacSystemFont, sans-serif;
  
  /* Code - readability */
  --md-code-font: "JetBrains Mono", "Fira Code", monospace;
}

/* Algorithm name styling */
.md-typeset h1 {
  font-weight: 700;
  letter-spacing: -0.02em;
  background: linear-gradient(135deg, #bd93f9 0%, #ff79c6 100%);
  -webkit-background-clip: text;
  -webkit-text-fill-color: transparent;
  background-clip: text;
}

/* Scientific paper-like section headers */
.md-typeset h2 {
  border-bottom: 2px solid #6272a4;
  padding-bottom: 0.5rem;
}

/* Reference/citation styling */
.md-typeset blockquote {
  border-left: 4px solid #bd93f9;
  background: rgba(98, 114, 164, 0.1);
  font-style: italic;
}

/* Math equation enhancement */
.md-typeset .arithmatex {
  overflow-x: auto;
  padding: 1rem;
  background: rgba(68, 71, 90, 0.5);
  border-radius: 4px;
}
```

---

### Phase 1: Core Setup (Week 1)
- [ ] Create `docs/` folder structure
- [ ] Configure `mkdocs.yml` with Dracula theme
- [ ] Add custom CSS (`docs/stylesheets/dracula-scientific.css`)
- [ ] Set up GitHub Actions for deployment

### Phase 2: Content Foundation (Week 2)
- [ ] Create landing page with project overview
- [ ] Write "Getting Started" guide
- [ ] Document installation methods
- [ ] Add basic API reference structure

### Phase 3: Algorithm Documentation (Weeks 3-4)
- [ ] Template for algorithm pages (scientific format)
- [ ] Document all 58 algorithms with:
  - Mathematical formulation (LaTeX)
  - Pseudocode
  - Parameters explanation
  - Usage examples
  - References (DOI links)

### Phase 4: Visualizations (Week 5)
- [ ] Implement `mkdocs-charts-plugin` with Vega-Lite
- [ ] Create JSON data structure for benchmarks
- [ ] Add convergence curves (JSON-based)
- [ ] Add comparison heatmaps
- [ ] Add box plots for stability analysis

### Phase 5: Advanced Features (Week 6)
- [ ] Mermaid diagrams for algorithm flowcharts
- [ ] Social cards with Dracula branding
- [ ] Search optimization
- [ ] Versioning setup
- [ ] Offline documentation build

---

## 📁 Proposed File Structure

```
docs/
├── index.md                          # Landing page
├── getting-started/
│   ├── installation.md
│   ├── quickstart.md
│   └── basic-concepts.md
├── algorithms/
│   ├── index.md                      # Algorithm overview/taxonomy
│   ├── swarm-intelligence/
│   │   ├── index.md
│   │   ├── particle-swarm.md
│   │   ├── ant-colony.md
│   │   └── ...
│   ├── evolutionary/
│   ├── gradient-based/
│   ├── classical/
│   ├── physics-inspired/
│   └── ...
├── benchmarks/
│   ├── index.md
│   ├── functions.md
│   └── comparison.md
├── api/
│   ├── index.md
│   └── reference.md
├── contributing/
│   └── guidelines.md
├── data/                             # JSON for Vega-Lite charts
│   ├── convergence/
│   ├── comparisons/
│   └── stability/
├── stylesheets/
│   └── dracula-scientific.css
└── assets/
    ├── logo.svg                      # Custom logo
    └── favicon.ico
```

---

## ⚙️ mkdocs.yml Configuration

```yaml
site_name: Useful Optimizer
site_url: https://anselmoo.github.io/useful-optimizer/
site_description: "58 optimization algorithms for numeric problems - A comprehensive Python library"
site_author: Anselmoo

repo_name: Anselmoo/useful-optimizer
repo_url: https://github.com/Anselmoo/useful-optimizer
edit_uri: edit/main/docs/

theme:
  name: material
  custom_dir: docs/overrides
  
  palette:
    # Dark mode (primary)
    - media: "(prefers-color-scheme: dark)"
      scheme: dracula
      toggle:
        icon: material/brightness-4
        name: Switch to light mode
    
    # Light mode
    - media: "(prefers-color-scheme: light)"
      scheme: dracula-light
      toggle:
        icon: material/brightness-7
        name: Switch to dark mode
  
  font:
    text: Inter
    code: JetBrains Mono
  
  features:
    - navigation.instant
    - navigation.instant.prefetch
    - navigation.tracking
    - navigation.tabs
    - navigation.tabs.sticky
    - navigation.sections
    - navigation.expand
    - navigation.path
    - navigation.indexes
    - navigation.top
    - search.suggest
    - search.highlight
    - search.share
    - content.tabs.link
    - content.code.copy
    - content.code.annotate
    - header.autohide
    - toc.follow
  
  icon:
    repo: fontawesome/brands/github
    logo: material/sine-wave
  
  favicon: assets/favicon.ico

extra_css:
  - stylesheets/dracula-scientific.css

extra_javascript:
  # Vega-Lite for charts
  - https://cdn.jsdelivr.net/npm/vega@5
  - https://cdn.jsdelivr.net/npm/vega-lite@5
  - https://cdn.jsdelivr.net/npm/vega-embed@6
  # MathJax for equations
  - javascripts/mathjax.js
  - https://unpkg.com/mathjax@3/es5/tex-mml-chtml.js

plugins:
  - search:
      lang: en
      separator: '[\s\-\_\.]+'
  - charts                            # mkdocs-charts-plugin
  - tags:
      tags_file: tags.md
  - social:                           # Auto-generate social cards
      cards_layout_options:
        background_color: "#282a36"
        color: "#f8f8f2"
  - minify:
      minify_html: true

markdown_extensions:
  # Python Markdown
  - abbr
  - admonition
  - attr_list
  - def_list
  - footnotes
  - md_in_html
  - tables
  - toc:
      permalink: true
      toc_depth: 3
  
  # PyMdown Extensions
  - pymdownx.arithmatex:
      generic: true
  - pymdownx.betterem
  - pymdownx.caret
  - pymdownx.details
  - pymdownx.emoji:
      emoji_index: !!python/name:material.extensions.emoji.twemoji
      emoji_generator: !!python/name:material.extensions.emoji.to_svg
  - pymdownx.highlight:
      anchor_linenums: true
      line_spans: __span
      pygments_lang_class: true
  - pymdownx.inlinehilite
  - pymdownx.keys
  - pymdownx.mark
  - pymdownx.smartsymbols
  - pymdownx.superfences:
      custom_fences:
        - name: mermaid
          class: mermaid
          format: !!python/name:pymdownx.superfences.fence_code_format
        - name: vegalite
          class: vegalite
          format: !!python/name:mkdocs_charts_plugin.fences.fence_vegalite
  - pymdownx.tabbed:
      alternate_style: true
  - pymdownx.tasklist:
      custom_checkbox: true
  - pymdownx.tilde

nav:
  - Home: index.md
  - Getting Started:
    - getting-started/index.md
    - Installation: getting-started/installation.md
    - Quick Start: getting-started/quickstart.md
    - Basic Concepts: getting-started/basic-concepts.md
  - Algorithms:
    - algorithms/index.md
    - Swarm Intelligence:
      - algorithms/swarm-intelligence/index.md
      - Particle Swarm: algorithms/swarm-intelligence/particle-swarm.md
      - Ant Colony: algorithms/swarm-intelligence/ant-colony.md
    - Evolutionary:
      - algorithms/evolutionary/index.md
    - Gradient-Based:
      - algorithms/gradient-based/index.md
    - Classical:
      - algorithms/classical/index.md
    - Physics-Inspired:
      - algorithms/physics-inspired/index.md
  - Benchmarks:
    - benchmarks/index.md
    - Test Functions: benchmarks/functions.md
    - Algorithm Comparison: benchmarks/comparison.md
  - API Reference: api/index.md
  - Contributing: contributing/guidelines.md

extra:
  status:
    new: Recently added
    deprecated: Deprecated
  social:
    - icon: fontawesome/brands/github
      link: https://github.com/Anselmoo/useful-optimizer
    - icon: fontawesome/brands/python
      link: https://pypi.org/project/useful-optimizer/
  analytics:
    provider: google
    property: G-XXXXXXXXXX  # Add your GA4 ID
  version:
    provider: mike
```

---

### Convergence Chart (No PNG Required)

**File:** `docs/data/convergence/particle_swarm.json`
```json
{
  "iterations": [0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100],
  "fitness": [15.2, 8.4, 4.1, 2.3, 1.2, 0.6, 0.3, 0.15, 0.08, 0.04, 0.02]
}
```

**Usage in Markdown:**
````markdown
```vegalite
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "title": "PSO Convergence on Ackley Function",
  "data": {"url": "data/convergence/particle_swarm.json"},
  "mark": {"type": "line", "point": true},
  "encoding": {
    "x": {"field": "iterations", "type": "quantitative", "title": "Iteration"},
    "y": {"field": "fitness", "type": "quantitative", "scale": {"type": "log"}, "title": "Best Fitness (log scale)"}
  },
  "config": {
    "background": "#282a36",
    "title": {"color": "#f8f8f2"},
    "axis": {"labelColor": "#f8f8f2", "titleColor": "#bd93f9", "gridColor": "#44475a"},
    "line": {"color": "#8be9fd"}
  }
}
```
````

---

## 🔬 Algorithm Page Template

```markdown
---
title: Particle Swarm Optimization
description: Social-inspired metaheuristic based on bird flocking behavior
icon: material/bird
status: new
tags:
  - swarm-intelligence
  - metaheuristic
  - population-based
---

# Particle Swarm Optimization

!!! abstract "Algorithm Summary"
    **Category:** Swarm Intelligence  
    **Year:** 1995  
    **Authors:** Kennedy & Eberhart  
    **Complexity:** $O(n \cdot p \cdot t)$

## Mathematical Formulation

The velocity and position updates are defined as:

$$
v_i^{t+1} = \omega v_i^t + c_1 r_1 (p_i - x_i^t) + c_2 r_2 (g - x_i^t)
$$

$$
x_i^{t+1} = x_i^t + v_i^{t+1}
$$

Where:

| Symbol | Description |
|--------|-------------|
| $v_i$ | Velocity of particle $i$ |
| $x_i$ | Position of particle $i$ |
| $\omega$ | Inertia weight |
| $c_1, c_2$ | Cognitive and social coefficients |
| $p_i$ | Personal best position |
| $g$ | Global best position |

## Usage

```python
from opt.swarm_intelligence import ParticleSwarm
from opt.benchmark.functions import shifted_ackley

optimizer = ParticleSwarm(
    func=shifted_ackley,
    lower_bound=-2.768,
    upper_bound=2.768,
    dim=2,
    population_size=30,
    max_iter=100,
)
best_solution, best_fitness = optimizer.search()
```

## Parameters

| Parameter | Type | Default | Description |
|-----------|------|---------|-------------|
| `population_size` | `int` | `30` | Number of particles |
| `max_iter` | `int` | `100` | Maximum iterations |
| `w` | `float` | `0.7` | Inertia weight |
| `c1` | `float` | `1.5` | Cognitive coefficient |
| `c2` | `float` | `1.5` | Social coefficient |

## Convergence Analysis

```vegalite
{
  "data": {"url": "../../data/convergence/particle_swarm.json"},
  "mark": "line",
  "encoding": {
    "x": {"field": "iteration", "type": "quantitative"},
    "y": {"field": "fitness", "type": "quantitative", "scale": {"type": "log"}}
  }
}
```

## References

!!! quote "Original Paper"
    Kennedy, J., & Eberhart, R. (1995). Particle swarm optimization.
    *Proceedings of ICNN'95*, 4, 1942-1948.
    [DOI: 10.1109/ICNN.1995.488968](https://doi.org/10.1109/ICNN.1995.488968)
```

---

## 📦 Dependencies

Add to `pyproject.toml`:

```toml
[project.optional-dependencies]
docs = [
    "mkdocs>=1.6.0",
    "mkdocs-material>=9.5.0",
    "mkdocs-charts-plugin>=0.0.10",
    "mkdocs-minify-plugin>=0.8.0",
    "pymdown-extensions>=10.7",
]
```

---

## 🚀 GitHub Actions Deployment

```yaml
# .github/workflows/docs.yml
name: Deploy Documentation

on:
  push:
    branches: [main]
    paths:
      - 'docs/**'
      - 'mkdocs.yml'
  workflow_dispatch:

permissions:
  contents: read
  pages: write
  id-token: write

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      
      - name: Setup Python
        uses: actions/setup-python@v5
        with:
          python-version: '3.12'
      
      - name: Install dependencies
        run: |
          pip install uv
          uv pip install --system mkdocs mkdocs-material mkdocs-charts-plugin mkdocs-minify-plugin pymdown-extensions
      
      - name: Build docs
        run: mkdocs build
      
      - name: Upload artifact
        uses: actions/upload-pages-artifact@v3
        with:
          path: site/

  deploy:
    needs: build
    runs-on: ubuntu-latest
    environment:
      name: github-pages
      url: ${{ steps.deployment.outputs.page_url }}
    steps:
      - name: Deploy to GitHub Pages
        id: deployment
        uses: actions/deploy-pages@v4
```

---

## 🎯 Success Criteria

- [ ] Documentation accessible at `anselmoo.github.io/useful-optimizer`
- [ ] All 58 algorithms documented with LaTeX formulas
- [ ] Interactive Vega-Lite charts rendering correctly
- [ ] Dark/light mode toggle working
- [ ] Lighthouse score > 90 for performance
- [ ] Mobile-responsive design
- [ ] Search functionality working across all pages

---

## Related Issues

- #39 - Generate benchmark study images (now JSON-based)
- #38 - Add doctest support (integrate with docs)

## Labels

`documentation`, `enhancement`, `design`
